### DISCUSSION TITLE
Měřící script neodesílá data

### DISCUSSION THREAD

**Discussion by Bajticek** [canarytrace/documentation#138] [RUM] [answered]
Ahoj,

napojil jsem pollyfill i měřící script dle popisu (viz obrázek):

![image](https://user-images.githubusercontent.com/29298855/204759893-6ff4f93d-aa61-4158-98b9-155337b1fa88.png)

Bohužel nenastává automatické odesílání posbíraných metrik na server, dle dokumentace: _https://rum.your-domain.com/rum - POST odesílání payloadu s metrikami_.

 Scripty se stáhnou ze serveru v pohodě:
![image](https://user-images.githubusercontent.com/29298855/204758378-47d8b26b-0a2f-4a83-b6ef-ed554f319ed3.png)

Jedná se o aplikaci v angular 13.

**Comment by rdpanek** [comment, marked as answer]
Ahoj @Bajticek co je v konzoli pri pouziti `?debug=1` ?

**Reply by Bajticek** [reply]
v konzoli bylo pusto prázdno... ale vypadá to na chybu při loadu scriptu z angularu. V případě statickeho přidání rovnou do .html nebo pridani staticky z assets, vše začne pracovat jak má a metriky se posílají. Děkuji za navedení pomocí debug modu. Diky němu se podařilo zprovoznit aspoň takto.

Pokud přidávám scripty dynamicky přes ng OnInit hook za pomoci setovani scriptu primo do domu, tak kód na měření metriky neběží. Zkusím ještě zjistit, jestli nemame chybu někde v posloupnosti bypassu scriptu nebo tak něco a případně bych dal vědět.

**Reply by rdpanek** [reply]
Ahoj @Bajticek vyřešeno?

**Reply by Bajticek** [reply]
ahoj, ano a ne, po přidání _debug=1_ se začli sbírat metriky do kibany. Bez tohoto parametru v url se však nic neodesílá.

**Reply by rdpanek** [reply]
`debug=1` nemá s odesíláním nic společného, pouze vytiskne připravený objekt. Můžeš stáhnout .har?

**Reply by Bajticek** [reply]
Tak problém se nakonec podařilo vyřešit. Debugem to opravdu nebylo, stále se jednalo o tentýž stav s importem scriptu. 
V případě že chceme injectnout script dynamicky přes JS:
```
    var e = document.querySelector('body').lastChild,
      s = document.createElement('script');
    s.setAttribute('type', 'text/javascript');
    s.setAttribute('src', 'https://rum.direct.cz/rum');
    s.setAttribute('id', 'canarytrace-rum');
    s.setAttribute('defer', '');
    e.parentNode.insertBefore(s, null);

  // to stejne pak do headu s pollyfilem
```
tak se ukládání metrik nepodaří. 

Pokud, ale vložím script staticky do html 
`<script src="https://rum.direct.cz/rum" id="canarytrace-rum" defer></script>`, tak jede jak má a metriky se odesílají.

Díky za pomoc. :-)

**Reply by rdpanek** [reply]
@Bajticek tak super 👌

**Reply by Bajticek** [reply]
jinak mám ještě otázku co se týká odesílání dat (volání api, které posílá sesbírané metriky). V jakém okamžiku by mělo docházet k jejich odesílání? Jde mi o to, že volání sužby nedochází při změně angulaří route (url). Vlastně se metrika odešle vlastně jen při načtení aplikace... Je toto chování správné? Případně je nějaká možnost jak vyvolat odeslání metrik i manuálně?

**Reply by rdpanek** [reply]
Tam se to v této verzi `1.10` odesílají minimálně dva payloady. První hlavní a druhý a někdy i třetí, když dojde k uvolnění metrik prohlížečem. FID je například závislé na akci uživatele. V připravovaném releasu https://github.com/canarytrace/documentation/issues/121 to bude fungovat jinak a to na základě nastaveného sampling rate. Manuální vyvolání by vám pomohlo jak, jaký by tam byl usecase?

**Reply by rdpanek** [reply]
asi mě napadá u SPA applikací? Kdy životní cyklus běží jen při prvním načtení, ale další flow už je řízeno klientským javascriptem?

**Reply by Bajticek** [reply]
přesně jak říkáte... při načtení dojde k odeslání payloudů (3x), ale potom už ne, protože se jedná o SPA a následující akce jsou řízené JS.

**Reply by rdpanek** [reply]
To dává smysl, přesunu to do nejbližšího release. Díky za hint.
